### PR TITLE
Project RSS feed

### PR DESCRIPTION
**Pull request by azmeuk**
Partially implements #1130, possibly `etags` and `if-modified-since` headers are missing, but let's discuss this in #1130

This patch generates a feed for each project, with a dynamic *secret* part in the URL based on the existing `Project.generate_token` method. The generated feed passes the [w3c validation](https://validator.w3.org/feed/check.cgi)

`<title>` and `<description>` can probably different, but the main data is here. Let me know what you prefer. Also, the content is simple enough so it does not need to be localized. If more information is needed and the feed content does need to be localized, maybe a lang variable can be added to the feed URL.

Maybe it worth adding an icon somewhere? Near the *Invite* button?  
A firefox extension like [awesome-rss](https://addons.mozilla.org/en-US/firefox/addon/awesome-rss/) automatically detects the feed on a project page, so I can live without it.

### CODE REVIEW

File: ihatemoney/templates/project_feed.xml
Comment (azmeuk) [inline ihatemoney/templates/project_feed.xml:15]: [`<author>`](https://www.rssboard.org/rss-specification#ltauthorgtSubelementOfLtitemgt) needs to be an email address, but [`<dc:creator>`](https://www.rssboard.org/rss-profile#namespace-elements-dublin-creator) can be anything.

**Comment by Glandos** [comment]
As mentioned:
> However in order to be able to do this, this would be needed to save the last update datetime for projects. This date would be updated at each bill addition/edition, and would be used to generate the ETag and Last-Modified headers.

We are using SQLAlchemy-continuum, so I guess that the last transaction is already accessible. But I don't have time to investigate it right now, it's just an idea.

**Comment by azmeuk** [comment]
@Glandos any chance you might have some time to review this soon :pray: ?
Do you have any questions about the implementation? If that can help, we can discuss on irc/matrix/whatever or have a call.

File: ihatemoney/models.py
Comment (almet) [inline ihatemoney/models.py:457]: You also need to change the `verify_token`docstring.

**Review by almet — COMMENTED** [review]
Thanks for working on this. It's working the right way to me, as the generated URLs will be able to only access the RSS endpoint.

Before merging, we should IMO decide if we want to integrate `etags` and `if-modified-since`, so that what we have in the main branch is feature-complete.

I advocate for the integration of these two =)

**Comment by almet** [comment]
> Maybe it worth adding an icon somewhere? Near the _Invite_ button?

What about adding a link and a button in the settings of the project?

File: ihatemoney/templates/project_feed.xml
Comment (azmeuk) [reply ihatemoney/templates/project_feed.xml:7]: I am not sure this is needed. Most readers display the channel name somewhere. Often you can also click select channel to only see the items for that specific channel. Displaying `I Hate Money` feels redundant to me.

File: ihatemoney/templates/project_feed.xml
Comment (azmeuk) [reply ihatemoney/templates/project_feed.xml]: I set *A simple shared budget manager web application*, is this OK?

**Comment by azmeuk** [comment]
> What about adding a link and a button in the settings of the project?

I added a `RSS` button next to the `CSS` and `JSON`. Not sure about the icon, I used pictures that were already there.

File: ihatemoney/templates/project_feed.xml
Comment (almet) [reply ihatemoney/templates/project_feed.xml]: LGTM!

File: MANIFEST.in
Comment (almet) [inline MANIFEST.in:2]: Good catch.

File: ihatemoney/templates/project_feed.xml
Comment (almet) [reply ihatemoney/templates/project_feed.xml:7]: I'm not sure I get it. Are you saying the channel name isn't coming from the `<title>`tag? If you only have the project name, how do you know it's coming from this software?

File: ihatemoney/tests/budget_test.py
Comment (almet) [inline ihatemoney/tests/budget_test.py:1741]: Well, actually it feels odd to me. You need to state that it's a feed with the latest bills from this project, maybe?

File: ihatemoney/tests/budget_test.py
Comment (almet) [inline ihatemoney/tests/budget_test.py:1773]: Just thinking about another test : we should probably also check that it's not possible to join the project with a "feed" token, I guess!

**Review by almet — COMMENTED** [review]
Looks good to me, I added a few other comments after a second pass. Hope you don't mind too much.

Thanks again for the work on this. I'm curious what other people think about etags and so on.

File: ihatemoney/templates/project_feed.xml
Comment (almet) [reply ihatemoney/templates/project_feed.xml]: Actually, not sure it's what describes best this feed. We should say it's the feed of the latest bills.

File: ihatemoney/templates/project_feed.xml
Comment (azmeuk) [reply ihatemoney/templates/project_feed.xml]: Ahh my bad. I though we were talking about the item title, and not the channel title. It makes sense, I fixed that.

File: ihatemoney/tests/budget_test.py
Comment (azmeuk) [reply ihatemoney/tests/budget_test.py:1784]: I have added that.

**Comment by azmeuk** [comment]
> Looks good to me, I added a few other comments after a second pass. Hope you don't mind too much.

Not at all, I understand what maintainership is and why you may be careful.

> I'm curious what other people think about etags and so on.

I'll look at this soon.

**Comment by azmeuk** [comment]
@almet to be able to return a `304` HTTP code, with `etags`/`if-modified-since`, we need to keep track of the *last update datetime* for a given project. I suppose this would mean to add a new column in the model, and update this whenever a bill is added/edited/deleted in the project.

Are you OK with that?

**Comment by almet** [comment]
> @almet to be able to return a `304` HTTP code, with `etags`/`if-modified-since`, we need to keep track of the _last update datetime_ for a given project. I suppose this would mean to add a new column in the model, and update this whenever a bill is added/edited/deleted in the project.
> 
> Are you OK with that?

As @Glandos noted earlier, we're using [SQLAlchemy-continuum](https://sqlalchemy-continuum.readthedocs.io/en/latest/) to version the changes, which should make things easier for us.

It took me some time to find out that [the `Transaction` class has a `issued_at` property](https://github.com/kvesteri/sqlalchemy-continuum/blob/master/sqlalchemy_continuum/transaction.py#L27) which might be useful for us here.

I'm not sure if we have to chose between `ETag`/`If-None-Match` and `if-modified-since`, or if we should do both of them? (seems redundant, but I guess it depends what's implemented in the clients?).

I advocate for using the easier one (unless there is something I don't get), which in this case seems to be `if-modified-since`.

**Comment by azmeuk** [comment]
Ok thank you for the tips I will look at that. I think both headers can be implemented at a cheap cost.

**Comment by azmeuk** [comment]
I just force-pushed the branch with the integration of the headers, but strangely github won't display the changes in the top tabs :man_shrugging: 

Anyways, the diff is available here : https://github.com/spiral-project/ihatemoney/commit/1c73d9fd1f1e31b7a75f5b0802579405bffc2156

I added pytest-libfaketime as a dev dependency to mock datetimes. I chose this instead of freezegun because, as freezegun is pure python it would not correctly mock sqlalchemy datetimes, but libfaketime works on a lower level and is able to mock sqlalchemy `issue_at` attributes (at least without a heavy setup).
libfaketime does not look very active, but this is because it is mainly stable. I maintain the pytest plugin, and in the past the maintainer was welcoming to contributions, so that wouldn't worry me.

I use continuum to find the last datetime edition for a project, and handle `if-modified-since`, and then I use this date to build a `etag` to manage `if-none-match`.

I could not track bills edition that way though, but addition/deletion works fine.

**Comment by azmeuk** [comment]
I just realized that I forgot to send the `etag` and `last-modified` headers back. I need to fix that.

**Comment by azmeuk** [comment]
This is better now.

File: tox.ini
Comment (azmeuk) [inline tox.ini:10]: `posargs` allows to pass arguments to pytest via tox, e.g. `tox -e py311 -- -v path/to/filetest.py`

Comment (zorun) [reply tox.ini:10]: Excellent, I wondered recently why I couldn't select a particular test this way.

File: ihatemoney/tests/budget_test.py
Comment (zorun) [inline ihatemoney/tests/budget_test.py:1]: Please add a test that checks that we can't access another project with a "feed" token. Our [first and only CVE](https://github.com/spiral-project/ihatemoney/security/advisories/GHSA-67j9-c52g-w2q9) was about unauthorized access by tweaking the project ID, so we definitely want to check that we don't introduce a similar issue for this new route (even if it's read-only).

Two cases are interesting: another project with the same private code, and another project with a different private code. Both should not allow access.

File: ihatemoney/models.py
Comment (zorun) [inline ihatemoney/models.py:481]: Add a test that checks that this token is indeed invalidated.

**Review by zorun — CHANGES_REQUESTED** [review]
I really like the approach, thanks a lot for working on this!

I left some comments, I hope you don't mind adding a few more tests ;)

File: ihatemoney/web.py
Comment (zorun) [reply ihatemoney/web.py:924]: Actually, since you only return the 100 most recent bills, you could also just check the last modified time of the 100 most recent bills. I suggest:

- fetch the 100 most recent bills (only once)
- use it to compute the last modified time
- use it again to feed it to the template (if needed according to the previous step)

File: ihatemoney/web.py
Comment (zorun) [inline ihatemoney/web.py:925]: Add a comment mentioning SQLAlchemy-Continuum, because it's not obvious from the code (and same thing above). If we ever decide to remove, upgrade or replace it, it will be easier to identify this particular usage.

By the way, will your code still work if a project has history disabled?

File: ihatemoney/web.py
Comment (azmeuk) [reply ihatemoney/web.py:927]: Done.

File: ihatemoney/tests/budget_test.py
Comment (azmeuk) [reply ihatemoney/tests/budget_test.py:1]: Done.

File: ihatemoney/models.py
Comment (azmeuk) [reply ihatemoney/models.py:481]: Done.

File: ihatemoney/templates/edit_project.html
Comment (almet) [reply ihatemoney/templates/edit_project.html:39]: That works for me as well, if you feel it's better.

File: ihatemoney/web.py
Comment (azmeuk) [reply ihatemoney/web.py:928]: > Add a comment mentioning SQLAlchemy-Continuum, because it's not obvious from the code (and same thing above). If we ever decide to remove, upgrade or replace it, it will be easier to identify this particular usage.

Done.

> By the way, will your code still work if a project has history disabled?

I added a test that checks that.

File: ihatemoney/templates/edit_project.html
Comment (azmeuk) [reply ihatemoney/templates/edit_project.html:39]: I added this in the dropdown menu, as the link to the feed is probably not something you want under your eyes at all time.
What do you think?

Comment (almet) [reply ihatemoney/templates/edit_project.html:39]: In the dropdown seems perfect.

Comment (almet) [reply ihatemoney/templates/edit_project.html:39]: If it's in the dropdown, I guess we won't need it in the export.

**Review by almet — APPROVED** [review]
LGTM!

File: ihatemoney/templates/edit_project.html
Comment (azmeuk) [reply ihatemoney/templates/edit_project.html]: Done.

File: ihatemoney/web.py
Comment (zorun) [reply ihatemoney/web.py:928]: Looks good, thanks!

File: ihatemoney/tests/budget_test.py
Comment (zorun) [reply ihatemoney/tests/budget_test.py:1]: I was thinking of accessing the RSS feed of another project with the RSS token. Your new tests try to use the RSS token as an invite token, which is already well covered by `test_invalid_invite_link_with_feed_token`.

Comment (azmeuk) [reply ihatemoney/tests/budget_test.py:1]: Ok that makes more sense :sweat_smile: 
I'll fix that.

**Comment by zorun** [comment]
Thanks for the changes, I still have a change request on the tests (testing we can't access other project's RSS feed), otherwise the rest looks good to me!

**Review by zorun — APPROVED** [review]
Nice work!

**Comment by almet** [comment]
Thanks again for the work on this. Yay!

**Comment by zorun** [comment]
@azmeuk we will do a release soon with this change (I just plan to finish the work on #1204 before that)

**Pull request merged**
